### PR TITLE
add default option for protocol setting

### PR DESCRIPTION
**Pull request by blacksponge**
## Purpose

This way the default settings on the GUI will use the default on the profile with the ability to override those settings.

## Context

This is a proposal following discussion on #217.

## How to test this PR

* Using the default setting on a dual stack backend should not change anything.
* Disabling one protocol should disable the tests using that protocol, could be checked in test results.
* Testing a domain using backend that only support one protocol (ex. zonemaster.ppsfleet.navy ) with the default settings should only trigger the tests of the supported protocol.

Screenshot 2021-06-14 at 16-36-46 Zonemaster

### CODE REVIEW

**Review by matsduf — CHANGES_REQUESTED** [review]
Your proposed update is not bad, and it follows my first suggestion. But I think we should give it a second thought. See my updated suggestion in https://github.com/zonemaster/zonemaster-gui/issues/217#issuecomment-859348817, which will also solve the original issue.

**Comment by blacksponge** [comment]
Updated version:
Screenshot 2021-06-14 at 16-36-46 Zonemaster

**Review by matsduf — COMMENTED** [review]
It looks fine. I will run a check before I formally approve it. Please add
```
"Disable IPv6": "Inaktivera IPv6",
"Disable IPv4": "Inaktivera IPv4"
```
in sv.json

File: src/environments/environment.ts [hunk 1]
```diff
@@ -6,6 +6,8 @@
 export const environment = {
   production: false,
   apiEndpoint: 'https://zonemaster.net/api',
+  //apiEndpoint: 'https://zonemaster.ppsfleet.navy/api',
```
Comment (matsduf) [inline src/environments/environment.ts:9]: What is it? What is it for? How and when is it used?

Comment (blacksponge) [reply src/environments/environment.ts:9]: Sorry, I pushed too much stuff, will revert.

**Review by matsduf — DISMISSED** [review]
This works as expected when tested. I approve this. Either we should have the translation in the remaining languages in this PR or in a separate one.

**Comment by matsduf** [comment]
@samisalmensuo, @repperille and @haarbo!

Screenshot 2021-06-14 at 16-36-46 Zonemaster

Could you provide the translations of "Disable IPv6" and "Disable IPv4", respectively, into Finnish, Norweigian and Danish, respectively? Just add a comment here.

**Comment by haarbo** [comment]
Disable is "Deaktiver" in Danish.

**Comment by matsduf** [comment]
> Disable is "Deaktiver" in Danish.

Thanks!

**Review by matsduf — DISMISSED** [review]

File: src/assets/i18n/fr.json [hunk 2]
```diff
@@ -48,5 +48,7 @@
 	"What is an undelegated domain test?": "C'est quoi un test sur un nom de domaine non délégué?",
 	"Zonemaster Backend is not available": "Zonemaster Backend n'est pas disponible",
 	"Profile": "Profil",
-	"Copyright © AFNIC and The Swedish Internet Foundation. All Rights Reserved": "© AFNIC et The Swedish Internet Foundation. Tous droits réservés"
+	"Copyright © AFNIC and The Swedish Internet Foundation. All Rights Reserved": "© AFNIC et The Swedish Internet Foundation. Tous droits réservés",
+	"Disable IPv6": "Désactivé IPv6",
+	"Disable IPv4": "Désactivé IPv4"
```
Comment (pnax) [inline src/assets/i18n/fr.json:53]: I would go with the infinitive form -> `Désactiver`.

**Review by pnax — COMMENTED** [review]
It looks like when a test finishes, the value is not properly reset.

Using the GUI from this PR (commit da6ebb3)
1. disable IPv4
2. run a test (for instance for afnic.fr)
3. wait for the test to finish and verify that IPv4 was disabled (OK)
4. open the « Options » menu again
5. both button are disabled
6. run a test (for instance for iis.se)
7. wait for the test to finish and see that IPv4 was disabled (not OK)

Either keep the button to disable IPv4 engaged or reset the value sent to the Backend.

File: src/assets/i18n/fr.json (hunk 2 above)
Comment (blacksponge) [reply src/assets/i18n/fr.json]: Fixed.

**Comment by blacksponge** [comment]
> It looks like when a test finishes, the value is not properly reset.

Good catch, I will fix that as well.

**Comment by matsduf** [comment]
I cannot build a release.
```
ERROR in src/app/components/form/form.component.html:13:33 - error TS2339: Property 'domain' does not exist on type '{}'.

13                    [(ngModel)]="form.domain"
                                   ~~~~~~~~~~~

  src/app/components/form/form.component.ts:13:16
    13   templateUrl: './form.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component FormComponent.
src/app/components/form/form.component.html:13:45 - error TS2339: Property 'domain' does not exist on type '{}'.

13                    [(ngModel)]="form.domain"
                                               
14                    (keypress)="($event.which === 13)? runDomainCheck() : 0"
   ~~~~~

  src/app/components/form/form.component.ts:13:16
    13   templateUrl: './form.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component FormComponent.
src/app/components/form/form.component.html:18:62 - error TS2339: Property 'domain' does not exist on type '{}'.

18               <p class="reset-form d-none d-sm-block" *ngIf="form.domain.length > 0 && !showProgressBar" >
                                                                ~~~~~~~~~~~

  src/app/components/form/form.component.ts:13:16
    13   templateUrl: './form.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component FormComponent.
```

**Comment by blacksponge** [comment]
Can you retry?

**Comment by matsduf** [comment]
> Can you retry?

Now I could build.

**Review by matsduf — DISMISSED** [review]
I have tested and I think it behaves as it should. This is an improvement over the current solution.

**Comment by matsduf** [comment]
The radio buttons here look wrong:
image

**Review by matsduf — CHANGES_REQUESTED** [review]
I suggest that the radio buttons about IPv4 and IPv6 are removed. In the report it says if any of them are disabled, and the buttons can only say it was requested to be disabled. It has no control of the profile. Most users of the GUI do not care about IPv4 and IPv6.

**Comment by blacksponge** [comment]
> The radio buttons here look wrong:

What do you mean?

Also, I think that the radio buttons in the result are beyond the scope of this MR.

**Comment by matsduf** [comment]
> Also, I think that the radio buttons in the result are beyond the scope of this MR.

This PR has changed how the radio buttons for IPv4 och IPv6 are displayed. The screen shot below shows how it looks like on zonemaster.net if you enable IPv4 and disable IPv6. IPv4 is green and IPv6 is red. With the updated code then the radio buttons are always red.

image

One solution is to remove them.

**Comment by blacksponge** [comment]
Oh okay, that probably comes from the fact that the ipv4 and ipv6 params are not explicitly stated in the request any more. Ideally the backend should send back the actual parameters that were used in the tests not just the one filled by the user (it will help on tests reusability too). But, yes, as a workaround they could be removed.

**Comment by blacksponge** [comment]
Acutally if I understand correctly, https://github.com/zonemaster/zonemaster-backend/pull/798 might solve that issue.

**Comment by matsduf** [comment]
> Acutally if I understand correctly, [zonemaster/zonemaster-backend#798](https://github.com/zonemaster/zonemaster-backend/pull/798) might solve that issue.

No, it has nothing to do with the Backend. Those buttons are set by the GUI based on the selection by the user.

**Comment by blacksponge** [comment]
Those button have nothing to do with the input form, they are set in the result component here
https://github.com/blacksponge/zonemaster-gui/blob/ip-option-default/src/app/components/result/result.component.html#L39
based on the data received from the `get_test_results` call here
https://github.com/blacksponge/zonemaster-gui/blob/ip-option-default/src/app/components/result/result.component.ts#L141 .

The `params` field in the result do not have any ipv4/ipv6 key as the input has not been normalized, hence resulting in `undefined` which is evaluated to false when interpreted in the template.

**Comment by matsduf** [comment]
> Those button have nothing to do with the input form, they are set in the result component here
(...)
> The `params` field in the result do not have any ipv4/ipv6 key as the input has not been normalized, hence resulting in `undefined` which is evaluated to false when interpreted in the template.

Thank you for clarifying. The change in this PR will make the ipv4/ipv6 keys being unset in the request. As far as I understand the value of the ipv4/ipv6 keys are copied from the request to the `get_test_results` response. It means that this FQ has created a bug concerning the ipv4/ipv6 buttons. I do not think it is acceptable to leave it as is (always being red), but I think it is acceptable to just remove the buttons because I do not think they add any value.

**Comment by mattias-p** [comment]
Aren't the markings there to signify what protocols were used for this particular test? I would think that's a good thing to know when you're trying to understand the results.

**Comment by matsduf** [comment]
If a protocol was disabled, there is a message in Basic. With this PR and the design of the Backend, the GUI has no access to the setting in the profile.

**Comment by matsduf** [comment]
image

**Comment by blacksponge** [comment]
>  As far as I understand the value of the ipv4/ipv6 keys are copied from the request to the get_test_results response

With zonemaster/zonemaster-backend#798 it seems that they will be set to the profile value if left unset. 
https://github.com/zonemaster/zonemaster-backend/pull/798/files#diff-7a0828c7a5501d04ebe0287191fe90c840c0cdb0a6670c77ff7e4ac257decec7R144-R145

**Comment by matsduf** [comment]
No, the method `generate_deterministic_hash` in  zonemaster/zonemaster-backend#798 will not change the value of the ipv4/ipv6 parameters, only create a local hash with effective values, and then create a hash. -- And that is good because that would change the behavior of the API.

**Comment by blacksponge** [comment]
But doesn't it also update the `encoded_params` hash that is then inserted into the db and is later the one retrieved by `test_results`? Or am I missing something?

**Comment by matsduf** [comment]
The deterministic_hash is a MD5 hash inserted in the database, but the function does not change the parameters themselves.

**Comment by blacksponge** [comment]
I am sorry but I don't understand how it doesn't change. From what I see,

* a new hash `to_encode` is created to compute the determinist hash [ref](https://github.com/pnax/zonemaster-backend/blob/deterministic-hash/lib/Zonemaster/Backend/DB.pm#L141),
* this hash is json encoded and returned to the caller along with the md5 hash [ref](https://github.com/pnax/zonemaster-backend/blob/deterministic-hash/lib/Zonemaster/Backend/DB.pm#L157)
* `create_new_test` use the returned encoded parameters [ref](https://github.com/pnax/zonemaster-backend/blob/deterministic-hash/lib/Zonemaster/Backend/DB/MySQL.pm#L129) to create the new test in the db [ref](https://github.com/pnax/zonemaster-backend/blob/deterministic-hash/lib/Zonemaster/Backend/DB/MySQL.pm#L158)
* those same parameters are used for the test  [ref](https://github.com/pnax/zonemaster-backend/blob/deterministic-hash/lib/Zonemaster/Backend/DB/MySQL.pm#L195) and return by `test_results` [ref](https://github.com/pnax/zonemaster-backend/blob/deterministic-hash/lib/Zonemaster/Backend/DB/MySQL.pm#L218)

The parameters are not modified in place but a new hash is replacing them to it is the same. We can probably discuss that at another place but I really feel like I am missing something because if it doesn't change I can't seem to see why.

**Comment by matsduf** [comment]
> I am sorry but I don't understand how it doesn't change. From what I see,

You are right, but that is just a draft that have not approved yet. Either we have to wait with this PR, or we have to resolve the bug within this PR. I suggest the latter.

**Comment by blacksponge** [comment]
Okay I'll do that next week then

**Comment by matsduf** [comment]
@mattias-p and @pnax!  Please express your opinion.

https://github.com/zonemaster/zonemaster-gui/pull/218/commits/86ab70a1227e38eb002783b66aba3e13b3f1cd39 removes the radio buttons that display of transport protocol (IPv4/IPv6) since with the update that will not be correct.

If we change the API  as suggested in  zonemaster/zonemaster-backend#798 then the buttons could be kept, but this PR cannot be merged until we have the change in the API.

Is displaying the transport protocol so important? In the messages you can always see when a protocol has been turned off. I do not think that most GUI users care about transport protocol. A think we can remove the radio buttons from the GUI (but the change in the API is maybe still good).

**Review by matsduf — APPROVED** [review]
This is fine. Can @mattias-p and @pnax review too?

**Comment by mattias-p** [comment]
I think displaying the query parameters on the result page is a great feature. Ideally we should have a clear section above the 
result messages where we display all the (normalized) query parameters in an structured way. That's out of scope for this PR, of course.

I don't think the result messages is a great place to communicate query parameters, but it's better than nothing. A couple of check boxes below a settings button is so great either. I don't mind having the check boxes removed, but then maybe we should also make an effort to get query parameters onto the result page in a more structured way.

File: e2e/FR12.e2e-spec.ts [hunk 3]
```diff
@@ -10,12 +10,12 @@ describe('Zonemaster test FR12 - [The simple view should support an advanced vie
...
-    expect(element(by.css('label[for="protocol_ipv6"]')).isPresent()).toBe(false);
+    expect(element(by.css('label[for="disable_protocol_ipv4"]')).isPresent()).toBe(false);
+    expect(element(by.css('label[for="disable_protocol_ipv6"]')).isPresent()).toBe(false);
   });
   it('should have [IPv4 checkbox] & [IPv6 checkbox] visible', () => {
     element(by.css('.switch')).click();
-    expect(element(by.css('label[for="protocol_ipv4"]')).isPresent()).toBe(true);
-    expect(element(by.css('label[for="protocol_ipv6"]')).isPresent()).toBe(true);
+    expect(element(by.css('label[for="disable_protocol_ipv4"]')).isPresent()).toBe(true);
+    expect(element(by.css('label[for="disable_protocol_ipv6"]')).isPresent()).toBe(true);
```
Comment (mattias-p) [inline e2e/FR12.e2e-spec.ts:19]: Maybe add "Disable" to the it() clauses?

File: src/app/components/form/form.component.html [hunk 4]
```diff
@@ -70,16 +70,15 @@ <h1>{{'Domain name'|translate}}</h1>
       <div *ngIf="is_advanced_options_enabled">
              <span [formGroup]="checkboxForm" class=" half-with">
               <span class="options d-flex" [formArrayName]="'items'" [class.invalid]="!checkboxForm['controls'].selectedItems.valid">
-                <span *ngFor="let protocol of checkboxForm['controls'].items['controls']; let i = index;"
+                <div *ngFor="let protocol of checkboxForm['controls'].items['controls']; let i = index;"
```
Comment (mattias-p) [inline src/app/components/form/form.component.html:73]: Is it proper nowadays to have div-tags inside of span-tags?

**Review by mattias-p — COMMENTED** [review]
Looks real good to me. I have a suggestion and a question.

File: src/app/components/form/form.component.html (hunk 4 above)
Comment (blacksponge) [reply src/app/components/form/form.component.html]: It is not ideal semantically speaking but would work without issue, I still revert it back to span

File: e2e/FR12.e2e-spec.ts (hunk 3 above)
Comment (blacksponge) [reply e2e/FR12.e2e-spec.ts]: This is done.

**Comment by pnax** [comment]
I agree with @mattias-p about having an easy way to see query parameters. This is fine for me to remove the buttons in the result page for now. But I think it would be nice to have them back when we'll decide to display the query parameters.

**Comment by blacksponge** [comment]
So can this be merged?

**Comment by matsduf** [comment]
> So can this be merged?

I will review.

**Review by pnax — COMMENTED** [review]
I tested it and it seems ok.
Before formally approving the PR, could you update the description with the right screenshot to avoid reading the comments and also add a small text on how to test this PR ?

**Comment by matsduf** [comment]
> @samisalmensuo, @repperille and @haarbo!
> 
> Screenshot 2021-06-14 at 16-36-46 Zonemaster
> 
> Could you provide the translations of "Disable IPv6" and "Disable IPv4", respectively, into Finnish, Norweigian and Danish, respectively? Just add a comment here.

@samisalmensuo and @repperille, could you provide the missing translation to Finnish and Norwegian, respectively?

**Review by matsduf — APPROVED** [review]
This looks fine. It behaves as expected when tested.

Having the translation into Finnish and Norwegian is not a show stopper for this PR, as it can be added by a separate PR afterwards. If the translations are not available when this is merged, please create an issue for that.

**Comment by matsduf** [comment]
@pnax and @mattias-p, please review.

**Comment by matsduf** [comment]
@blacksponge, I think you can merge.

**Review by pnax — APPROVED** [review]

**Comment by mattias-p** [comment]
I've tested this for v2021.2 and it LGTM.

**Pull request merged**